src/app-admin.js

```diff
@@ -0,0 +1,491 @@
+  if (!session) return;
+
+  if (!checkFilePermission(fileId, session.user_id, res)) return;
+
+  const users = getAccountDb().all(
+    `SELECT users.id as userId, user_name as userName, display_name as displayName, 
+            CASE WHEN user_access.file_id IS NULL THEN 0 ELSE 1 END as haveAccess,
+            CASE WHEN files.id IS NULL THEN 0 ELSE 1 END as owner
+      FROM users
+      LEFT JOIN user_access ON user_access.file_id = ? and user_access.user_id = users.id
+      LEFT JOIN files ON files.id = ? and files.owner = users.id
+      WHERE users.enabled = 1 AND users.user_name <> ''`,
+    [fileId, fileId],
+  );
+  res.json(users);
+});
+
+app.post('/access/get-bulk', (req, res) => {
+  const fileIds = req.body || {};
+  const accessMap = new Map();
+
+  fileIds.forEach((fileId) => {
+    const userAccess = getAccountDb().all(
+      `SELECT user_access.file_id as fileId, user_access.user_id as userId, users.display_name as displayName, users.user_name as userName
+       FROM users
+       JOIN user_access ON user_access.user_id = users.id
+       WHERE user_access.file_id = ?
+       UNION
+       SELECT files.id, users.id, users.display_name, users.user_name
+       FROM users
+       JOIN files ON files.owner = users.id
+       WHERE files.id = ?`,
+      [fileId, fileId],
+    );
+    accessMap.set(fileId, userAccess);
+  });
+
+  res.status(200).send({ status: 'ok', data: Array.from(accessMap.entries()) });
+});
+
+app.get('/access/check-access', (req, res) => {
+  const fileId = req.query.fileId;
+  const session = validateUser(req, res);
+  if (!session) return;
+
+  const { id: fileIdInDb } = getFileById(fileId);
+  if (!fileIdInDb) {
+    sendErrorResponse(res, 400, 'invalid-file-id', 'File not found at server');
+    return false;
+  }
+
+  if (isAdmin(session.user_id)) {
+    res.json({ granted: true });
+    return;
+  }
+
+  const { owner } =
+    getAccountDb().first(
+      `SELECT files.owner
+     FROM files
+     WHERE files.id = ?`,
+      [fileId],
+    ) || {};
+
+  res.json({ granted: owner === session.user_id });
+});
+
+app.post('/access/transfer-ownership/', (req, res) => {
+  const newUserOwner = req.body || {};
+  const session = validateUser(req, res);
+  if (!session) return;
+
+  if (!checkFilePermission(newUserOwner.fileId, session.user_id, res)) return;
+
+  if (!newUserOwner.newUserId) {
+    sendErrorResponse(res, 400, 'user-cant-be-empty', 'User cannot be empty');
+    return;
+  }
+
+  const { cnt } =
+    getAccountDb().first(
+      'SELECT count(*) AS cnt FROM users WHERE users.id = ?',
+      [newUserOwner.newUserId],
+    ) || {};
+
+  if (cnt === 0) {
+    sendErrorResponse(res, 400, 'new-user-not-found', 'New user not found');
+    return;
+  }
+
+  getAccountDb().mutate('UPDATE files SET owner = ? WHERE id = ?', [
+    newUserOwner.newUserId,
+    newUserOwner.fileId,
+  ]);
+
+  res.status(200).send({ status: 'ok', data: {} });
+});
+
+app.get('/file/owner', async (req, res) => {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏is this used to get the owner of a specific file? Could we attach it to the endpoint that serves the list of available files?

Reply:
this is important to return not only the id, but the username / displayname so we can show it in the file list. it can be baked into the file list. i will do it

Reply:
removed it in favor of use the modified /list-user-files